com.unity.shadergraph/Editor/ShaderFoundry/Utilities/VirtualTextureAttributes.cs

```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Collections.Generic;
+
+namespace UnityEditor.ShaderFoundry
+{
+    internal class VirtualTextureAttribute
+    {
+        const string AttributeName = "VirtualTexture";
+        const string LayerCountParamName = "layerCount";
+
+        internal const int MaxLayerCount = 4;
+        internal int LayerCount = 2;
+
+        internal ShaderAttribute Build(ShaderContainer container)
+        {
+            var attributeBuilder = new ShaderAttribute.Builder(container, AttributeName);
+            attributeBuilder.Param(new ShaderAttributeParam.Builder(container, null, LayerCount.ToString()).Build());
+            return attributeBuilder.Build();
+        }
+
+        internal static VirtualTextureAttribute FindFirst(IEnumerable<ShaderAttribute> attributes)
+        {
+            var attribute = attributes.FindFirst(AttributeName);
+            if (attribute.IsValid)
+                return TryParse(attribute);
+            return null;
+        }
+
+        internal static VirtualTextureAttribute TryParse(ShaderAttribute attribute)
+        {
+            if (attribute.Name != AttributeName)
+                return null;
+
+            var result = new VirtualTextureAttribute();
+            var index = 0;
+            foreach (var param in attribute.Parameters)
+            {
+                if (index > 0)
```

Comment:
Why not check it outside of the loop?
Is there a Parameters.Count or .Length or something similar?

Reply:
The parameters are an enumerable which you can't get the size of without effectively converting it into a list or iterating it once.